plugins/destination/bigquery/client/read.go

```diff
@@ -97,20 +36,153 @@ func (c *Client) Read(ctx context.Context, table *schema.Table, sourceName strin
 	if err != nil {
 		return fmt.Errorf("failed to read table %s: %w", table.Name, err)
 	}
+	arrowSchema := table.ToArrowSchema()
 	for {
-		values := c.createResultsArray(table)
-		v := make([]any, len(values))
+		values := make([]bigquery.Value, len(table.Columns))
 		err := it.Next(&values)
 		if err == iterator.Done {
 			break
 		}
 		if err != nil {
 			return fmt.Errorf("failed to read from table %s: %w", table.Name, err)
 		}
+		rb := array.NewRecordBuilder(memory.DefaultAllocator, arrowSchema)
 		for i := range values {
-			v[i] = values[i]
+			readItem(values, rb, i)
 		}
-		res <- v
+		res <- rb.NewRecord()
 	}
 	return nil
 }
+
+func readItem(values []bigquery.Value, rb *array.RecordBuilder, i int) error {
+	fieldBuilder := rb.Field(i)
+	value := values[i]
+	return appendValue(fieldBuilder, value)
+}
+
+func appendValue(bldr array.Builder, value any) error {
+	if value == nil {
+		bldr.AppendNull()
+		return nil
+	}
+	switch arr := bldr.(type) {
```

Comment:
```suggestion
	switch bldr := bldr.(type) {
```
could improve readability, as `arr` suggests that we have `arrow.Array` rather then `array.Builder`